### PR TITLE
Fix crash when using adaptivity with tri/tet elements in porous flow

### LINKED ISSUE #14428: PorousFlow: Stateful props + adaptivity + TET4/TRI3 elements = crash
## Bug Description

*Update 2* I think the stateful/adaptivity and TET4/TRI3 mesh issue is different to this, so I'll make a new issue, and keep this one for the porous flow specific crash.

~~*Update*: Looks like the combination of stateful properties, adaptivity and TET4/TRI3 element type causes an issue in general, not just in porous flow.~~

Mesh adaptivity with a tet4 mesh (in 3D) or a tri3 mesh (in 2D) causes a segfault in PorousFlow, as discovered by Daniel, see mailing list discussion https://groups.google.com/forum/#!topic/moose-users/6yUh7_WCGHA

## Steps to Reproduce

~~*Update*: change the element type to TET4 in test/tests/materials/stateful_prop/stateful_prop_adaptivity_test.i~~

The following simple input file illustrates this
```
[Mesh]
  [./mesh]
    type = GeneratedMeshGenerator
    elem_type = TRI3
    dim = 2
    # elem_type = TET4
    # dim = 3
    nx = 1
  [../]
[]

[Adaptivity]
  marker = marker
  [./Markers]
    [./marker]
      type = ValueRangeMarker
      variable = pp
      lower_bound = 0.1
      upper_bound = 1
    [../]
  [../]
[]

[GlobalParams]
  PorousFlowDictator = dictator
[]

[Variables]
  [./pp]
    initial_condition = 0
  [../]
[]

[Kernels]
  [./mass]
    type = PorousFlowMassTimeDerivative
    variable = pp
  [../]
[]

[BCs]
  [./left]
    type = PresetBC
    variable = pp
    boundary = left
    value = 1
  [../]
  [./right]
    type = PresetBC
    variable = pp
    boundary = right
    value = 0
  [../]
[]

[UserObjects]
  [./dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pp'
    number_fluid_phases = 1
    number_fluid_components = 1
  [../]
  [./pc]
    type = PorousFlowCapillaryPressureVG
    m = 0.5
    alpha = 1
  [../]
[]

[Modules]
  [./FluidProperties]
    [./simple_fluid]
      type = SimpleFluidProperties
      bulk_modulus = 1.2
      density0 = 1
      viscosity = 1
      thermal_expansion = 0
    [../]
  [../]
[]

[Materials]
  [./temperature]
    type = PorousFlowTemperature
  [../]
  [./ppss]
    type = PorousFlow1PhaseP
    porepressure = pp
    capillary_pressure = pc
  [../]
  [./massfrac]
    type = PorousFlowMassFraction
  [../]
  [./simple_fluid]
    type = PorousFlowSingleComponentFluid
    fp = simple_fluid
    phase = 0
  [../]
  [./porosity]
    type = PorousFlowPorosityConst
    porosity = 0.1
  [../]
[]

[Postprocessors]
  [./numdofs]
    type = NumDOFs
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  end_time = 3
  dt = 1
  solve_type = Newton
  nl_abs_tol = 1e-12
[]

[Outputs]
  execute_on = 'timestep_end'
[]
```

## Impact
Can't use adaptivity with tetrahedral meshes

### PR DESCRIPTION
**Pull request by cpgr**
Previously, using tri or tet elements with mesh adaptivity would
cause a seg fault with kernels that use stateful material properties
evaluated at nodes.

Closes #14428

### CODE REVIEW

**Comment by cpgr** [comment]
@WilkAndy - I think I have finally got to the bottom of #14428, but I'm not sure this is the perfect fix, although it seems to work ok in my limited testing.

The problem was the number of nodes is less than the number of qps, so each nodal material has an empty value in the storage for each node (as properties were only computed for the number of nodes for these nodal materials).

When mesh adaptivity takes place, stateful properties are populated for the new child elements using the closest qps in the parent element. Unfortunately in this case, the empty qp in the material property can be propagated to the old and older properties for a child element. When old nodal material properties are requested, an empty vector may be present, so a seg fault occurs when trying to get properties from the vector. 

To fix this, I've added some code to fill the empty values by copying from the previous node, so that an empty vector is never added to the old/older values for a child element. In theory, it should be possible to work out which of the properties to copy using the known map between parent and child elements, but I haven't been able to get this to work.

**Comment by WilkAndy** [comment]
Wow, this is good stuff.  It must have taken you ages to do it.

**Comment by cpgr** [comment]
I've been trying to close off all my old tickets, so have spent days thinking about what has been going wrong while doing some actual work. I'm still not sure what is the best way to fix this, but at least it doesn't crash anymore!

**Comment by loganharbour** [comment]
What's the status of this?

File: modules/porous_flow/include/materials/PorousFlowMaterial.h [hunk 1]
```diff
@@ -73,4 +79,6 @@ class PorousFlowMaterial : public Material
   const VariableName _saturation_variable_name;
   const VariableName _temperature_variable_name;
   const VariableName _mass_fraction_variable_name;
+
+  std::vector<unsigned int> _qpmap{{0, 1, 2, 3}};
```
Comment (loganharbour) [inline modules/porous_flow/include/materials/PorousFlowMaterial.h:83]:
```suggestion
```

**Comment by WilkAndy** [comment]
Yea, this should be merged if it works, right, @cpgr ?

File: modules/porous_flow/include/materials/PorousFlowMaterial.h (hunk 1 above)
Comment (cpgr) [reply modules/porous_flow/include/materials/PorousFlowMaterial.h:83]: Oops - forgot to delete this after using it for some debugging

**Comment by cpgr** [comment]
I think that this is the best I can do for this issue, in that it fixes the crashing when using test/tris and mesh adaptivity and seems to produce reasonable results when used. The better fix would be a nodal material class, but that is a lot of work!

Do either of you know Daniel Birdsell's GitHub username? I think he originally discovered this problem so he might be interested in trying it.

**Comment by WilkAndy** [comment]
Sorry, i don't know DB's github username

**Review by loganharbour — APPROVED** [review]

**Pull request merged**